
syntax = "proto3";
package mini2;

message NodeId { string id = 1; }
message Heartbeat { string from = 1; int64 ts_unix_ms = 2; }
message HeartbeatAck { bool ok = 1; }

message Request {
  string request_id = 1;
  string query = 2;
  bool need_green = 3;
  bool need_pink = 4;
}

message WorkerResult {
  string request_id = 1;
  uint32 part_index = 2;
  bytes payload = 3;
}

message AggregatedResult {
  string request_id = 1;
  uint64 total_rows = 2;
  uint64 total_bytes = 3;
  repeated bytes chunks = 4; // Strategy A only
}

message SessionOpen { 
  string request_id = 1; 
  bool accepted = 2;  // Whether request was accepted
  string status = 3;  // "QUEUED", "PROCESSING", "REJECTED"
  int64 timestamp_ms = 4;  // When request was accepted
}

message NextChunkReq { string request_id = 1; uint32 next_index = 2; }
message NextChunkResp { string request_id = 1; bool has_more = 2; bytes chunk = 3; }
message PollReq { string request_id = 1; }
message PollResp { string request_id = 1; bool ready = 2; bytes chunk = 3; bool has_more = 4; }

message CloseSessionReq { string session_id = 1; }
message CloseSessionResp { bool success = 1; }

// Broadcast and Control messages
message BroadcastMessage {
  string message_type = 1;  // "shutdown", "status", "health_check"
  string from_node = 2;
  bytes payload = 3;
}

message ShutdownRequest {
  string from_node = 1;
  int32 delay_seconds = 2;  // Graceful shutdown delay
}

message ShutdownResponse {
  bool acknowledged = 1;
  string node_id = 2;
}

message StatusRequest {
  string from_node = 1;
}

message StatusResponse {
  string node_id = 1;
  string state = 2;  // "IDLE", "BUSY", "OVERLOADED"
  int32 queue_size = 3;
  int64 uptime_seconds = 4;
  int32 requests_processed = 5;
  uint64 memory_bytes = 6;  // Current memory usage in bytes
}

service NodeControl {
  rpc Ping(Heartbeat) returns (HeartbeatAck);
  rpc Broadcast(BroadcastMessage) returns (HeartbeatAck);
  rpc Shutdown(ShutdownRequest) returns (ShutdownResponse);
  rpc GetStatus(StatusRequest) returns (StatusResponse);
}

service TeamIngress {
  rpc HandleRequest(Request) returns (HeartbeatAck);
  rpc PushWorkerResult(WorkerResult) returns (HeartbeatAck);
}

service ClientGateway {
  rpc OpenSession(SessionOpen) returns (HeartbeatAck);
  rpc GetNext(NextChunkReq) returns (NextChunkResp);
  rpc StartRequest(Request) returns (SessionOpen);
  rpc PollNext(PollReq) returns (PollResp);
  rpc CloseSession(CloseSessionReq) returns (CloseSessionResp);
}
